File: core/src/com/xenoage/zong/core/music/chord/Note.java
package com.xenoage.zong.core.music.chord;

import com.xenoage.utils.annotations.Const;
import com.xenoage.utils.annotations.NonNull;
import com.xenoage.zong.core.music.Pitch;
import lombok.Data;
import lombok.val;

import java.util.ArrayList;


/**
 * Class for a single note.
 * Within a score, it is always part of a chord.
 *
 * @author Andreas Wenger
 */
@Const @Data public final class Note {

	@NonNull private final Pitch pitch;


	/**
	 * Returns a list of {@link Note}s from the given {@link Pitch}es.
	 */
	public static ArrayList<Note> notes(Pitch... pitches) {
		ArrayList<Note> ret = new ArrayList<>(pitches.length);
		for (val p : pitches)
			ret.add(new Note(p));
		return ret;
	}


	@Override public String toString() {
		return "note(" + pitch.toString() + ")";
	}


}

File: core/src/com/xenoage/zong/core/music/chord/Note.kt
package com.xenoage.zong.core.music.chord

import com.xenoage.zong.core.music.Pitch


/**
 * Class for a single note.
 * Within a score, it is always part of a chord.
 */
data class Note(
	val pitch: Pitch
) : Comparable<Note> {

	override fun toString() = "note($pitch)"

	override fun compareTo(other: Note) = pitch.compareTo(other.pitch)

}

/**
 * Returns a list of [Note]s from the given [Pitch]es.
 */
fun notes(vararg pitches: Pitch) = pitches.map(::Note)
